dropshot/src/handler.rs

```diff
@@ -806,43 +856,83 @@ impl<T: JsonSchema + Serialize + Send + Sync + 'static>
     From<HttpResponseOkObject<T>> for HttpHandlerResult
 {
     fn from(response: HttpResponseOkObject<T>) -> HttpHandlerResult {
-        response.for_object(&response.0)
+        HttpResponseOkObject::for_object(&response.0)
     }
 }
 
 /**
- * `HttpResponseOkObjectList<T: Serialize>` wraps a collection of serializable
- * types.  It denotes an HTTP 200 "OK" response whose body is generated by
- * serializing the sequence of objects.
- * TODO-polish We will probably want to add headers for the total result set
- * size and the number of results that we're returning here, plus the marker.
- * TODO-cleanup move/copy the type aliases from src/api_model.rs?
- */
-pub struct HttpResponseOkObjectList<T: Serialize + Send + Sync + 'static>(
-    pub Vec<T>,
-);
-impl<T: JsonSchema + Serialize + Send + Sync + 'static> HttpTypedResponse
-    for HttpResponseOkObjectList<T>
+ * Wraps a list of items with a response intended for pagination.  See the
+ * pagination documentation docs for more details.
+ */
+pub struct HttpResponseOkPage<I>(ClientPage<I>)
+where
+    I: JsonSchema + Serialize + Send + Sync + 'static;
+
+impl<I> HttpTypedResponse for HttpResponseOkPage<I>
+where
+    I: JsonSchema + Serialize + Send + Sync + 'static,
 {
-    type Body = Vec<T>;
+    type Body = ClientPage<I>;
     const STATUS_CODE: StatusCode = StatusCode::OK;
 }
-impl<T: JsonSchema + Serialize + Send + Sync + 'static>
-    From<HttpResponseOkObjectList<T>> for HttpHandlerResult
+
+impl<I> HttpResponseOkPage<I>
+where
+    I: JsonSchema + Serialize + Send + Sync + 'static,
 {
-    fn from(list_wrap: HttpResponseOkObjectList<T>) -> HttpHandlerResult {
-        let list = list_wrap.0;
-        let buffer_list = list.iter().map(serialize_json_stream_element);
-        let mut bytebuf = BytesMut::new();
-        for maybe_buffer in buffer_list {
-            let buffer = maybe_buffer?;
-            bytebuf.put(buffer);
-        }
+    pub fn new_for_marker<F, S>(
+        items: Vec<I>,
+        get_field_value: F,
+    ) -> Result<HttpResponseOkPage<I>, HttpError>
+    where
+        F: Fn(&I) -> S,
+        S: Serialize + Send + Sync + 'static,
+    {
+        let next_page = items
+            .last()
+            .map(|last_item| {
+                let field_value = get_field_value(last_item);
+                let selector = MarkerPageSelector::Marker(field_value);
+                PageToken::new(selector).to_serialized()
+            })
+            .transpose()?;
+
+        Ok(HttpResponseOkPage(ClientPage {
+            next_page,
+            items,
+        }))
+    }
 
-        Ok(Response::builder()
-            .status(HttpResponseOkObjectList::<T>::STATUS_CODE)
-            .header(http::header::CONTENT_TYPE, CONTENT_TYPE_NDJSON)
-            .body(bytebuf.freeze().into())?)
+    pub fn new_with_paginator<F, ScanMode, PageSelector>(
+        items: Vec<I>,
```

Comment:
Yeah, it'd be neat to just accept an iterator, although we might need to `::collect()` it into a Vec in order to serialize it.

I like the idea of putting the `take(limit)` in here, but I think the most common use case is probably that the list endpoint is backed by a database query and the consumer will want to put the limit into the SQL query.  Given that, is it worth trying to put this in here?  I'm not sure.  There is a question of how we want to handle the case where the caller provided more items than the limit.  We could panic with an explicit message, or produce a 500 server error, or silently drop the extra items.  If we put the `take(limit)` in here, then on the plus side, it will "just work" for examples like the ones in the repo and it won't be wrong for consumers that have already applied the limit by the time we get here.  On the other hand, it might be pretty confusing behavior that would be pretty tough to debug.

Reply:
That's a good point that the SQL query is probably the right place to apply the limit. I agree that we should leave it as-is, and maybe keep an eye out as we build consumers for common patterns that we can simplify.

Reply:
I've deferred the question of enforcing the limit.  I will file an issue for it so we don't forget about it.